client/polling/poll.go

```diff
@@ -0,0 +1,154 @@
+package polling
+
+import (
+	"context"
+	"errors"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/rpc/client"
+	coretypes "github.com/tendermint/tendermint/rpc/core/types"
+	"math"
+	"math/rand"
+	"strings"
+	"time"
+)
+
+const (
+	magicForBlockDoesNotExistErr = "must be less than or equal to the current blockchain height"
+	baseDelayBetweenTransactions = 25 * time.Millisecond
+	maxDelayBetweenTransactions  = 2500 * time.Millisecond
+)
+
+var (
+	errCapacityInvalid = errors.New("capacity must be 1 or greater")
+	errNoBlocks        = errors.New("rpc call for latest block returned nil")
+)
+
+func PollForBlocks(ctx context.Context, logger log.Logger, c client.Client, outCapacity ...int) (out <-chan abci.ResponseDeliverTx, err error) {
+	logger = logger.With("cmp", "block-poller")
+
+	// Create an outgoing channel for events
+	outCap := 1
+	if len(outCapacity) > 0 {
+		outCap = outCapacity[0]
+		if outCap < 1 {
+			return nil, errCapacityInvalid
+		}
+	}
+	outc := make(chan abci.ResponseDeliverTx, outCap)
+
+	// Get the current height
+	block, err := c.Block(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if block == nil || block.Block == nil {
+		return nil, errNoBlocks
+	}
+
+	height := block.Block.Height
+
+	// Run until context dies
+	go func() {
+		defer close(outc)
+		err := pollForever(ctx, logger, c, height, outc)
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			return
+		}
+		logger.Error("failed polling for events", "err", err)
+	}()
+	return outc, nil
+}
+
+
+
+func pollForever(ctx context.Context, logger log.Logger, c client.Client, startHeight int64, output chan<- abci.ResponseDeliverTx) error {
+	const extraDelay = 100.0 // milliseconds
+	const blockPeriod = time.Second * 5
+	currentHeight := startHeight
+	requiredDelay := blockPeriod
+
+	// compute the sliding average of the time between blocks, starting at double the block period
+	avg := newSlidingAverage(2*blockPeriod.Seconds(), 20)
+
+	var lastBlock *coretypes.ResultBlock
+
+	for {
+		if requiredDelay > 0 {
+			delay := time.After(requiredDelay)
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-delay:
+			}
+		}
+
+		block, err := c.Block(ctx, &currentHeight)
+		if err != nil {
+			// Check to see if the block was requested too soon
+			if strings.Contains(err.Error(), magicForBlockDoesNotExistErr) {
+				// Wait a little bit longer, then ask for the block again
+				requiredDelay = blockPeriod / 4
+				requiredDelay += time.Duration(rand.Intn(250)) * time.Millisecond
+				continue
+			}
+			logger.Error("could not retrieve block", "height", currentHeight, "err", err)
+
+			requiredDelay = blockPeriod * 2
```

Comment:
why 2x block period?

Reply:
I just figured we'd start at a higher number then move downwards to the "real" value as the first 20 blocks come in